src/screen-cast.c

```diff
@@ -373,12 +527,134 @@ validate_cursor_mode (const char *key,
   return TRUE;
 }
 
+static gboolean
+validate_restore_token (const char *key,
+                        GVariant *value,
+                        GVariant *options,
+                        GError **error)
+{
+  const char *restore_token = g_variant_get_string (value, NULL);
+
+  if (!g_uuid_string_is_valid (restore_token))
+    {
+      g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Restore token is not a valid UUID string");
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
+static gboolean
+validate_persist_mode (const char *key,
+                       GVariant *value,
+                       GVariant *options,
+                       GError **error)
+{
+  uint32_t mode = g_variant_get_uint32 (value);
+
+  if (__builtin_popcount (mode) != 1)
+    {
+      g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                   "Invalid persist mode %x", mode);
+      return FALSE;
+    }
+
+  return TRUE;
+}
+
 static XdpOptionKey screen_cast_select_sources_options[] = {
   { "types", G_VARIANT_TYPE_UINT32, validate_device_types },
   { "multiple", G_VARIANT_TYPE_BOOLEAN, NULL },
   { "cursor_mode", G_VARIANT_TYPE_UINT32, validate_cursor_mode },
+  { "restore_token", G_VARIANT_TYPE_STRING, validate_restore_token },
+  { "persist_mode", G_VARIANT_TYPE_UINT32, validate_persist_mode },
 };
 
+static gboolean
+replace_restore_token_with_data (ScreenCastSession *screen_cast_session,
+                                 GVariant **in_out_options,
+                                 GError **error)
+{
+  Session *session = (Session *)screen_cast_session;
+  GVariantBuilder options_builder;
+  g_autoptr(GVariant) options = NULL;
+  gsize i;
+
+  options = *in_out_options;
+
+  g_variant_lookup (options, "persist_mode", "u", &screen_cast_session->persist_mode);
+
+  g_variant_builder_init (&options_builder, G_VARIANT_TYPE_VARDICT);
+  for (i = 0; i < G_N_ELEMENTS (screen_cast_select_sources_options); i++)
+    {
+      g_autoptr(GVariant) value = NULL;
+
+      value = g_variant_lookup_value (options,
+                                      screen_cast_select_sources_options[i].key,
+                                      screen_cast_select_sources_options[i].type);
+
+      if (!value)
+        continue;
+
+      if (g_strcmp0 (screen_cast_select_sources_options[i].key, "restore_token") == 0)
+        {
+          g_autoptr(GVariant) restore_data = NULL;
+          g_autofree char *restore_token = NULL;
+
+          if (is_remote_desktop_session (session))
+            {
+              g_set_error (error, XDG_DESKTOP_PORTAL_ERROR, XDG_DESKTOP_PORTAL_ERROR_INVALID_ARGUMENT,
+                           "Remote desktop sessions cannot be restored");
```

Comment:
I assumed the fact that this is not documented in `org.freedesktop.portal.RemoteDesktop` would be enough :slightly_smiling_face: 

Do you have any particular suggestions as to where to document this?

Reply:
What I mean is that this code cannot be reached by using `org.freedesktop.portal.RemoteDesktop` - it will bail out earlier, when parsing the options and noticing that `persist_mode` is there but not allowed. I added this check in the code merely as a security measure, I don't expect that this code is ever reached in practice.

Reply:
This function is called from `org.freedesktop.ScreenCast.SelectSources`, which can be called on a remote desktop session too. Or am I missing somewhere where we already return an error before getting here?

Reply:
Oh, I understand what you mean now! Sorry for this confusion. I've update this pull request to (1) fail earlier if it's a remote desktop session and has persist_mode != 0, and (2) document this behavior in the XML file.